File: src/main/java/io/chesslave/visual/SquareImage.java
package io.chesslave.visual;

import io.chesslave.model.Board;
import io.chesslave.model.Square;
import javaslang.Lazy;
import java.awt.image.BufferedImage;

public final class SquareImage {

    private final Lazy<BufferedImage> image;
    private final Square square;
    private final boolean flipped;
    private final int size;

    public SquareImage(BufferedImage board, Square square, boolean flipped) {
        this.square = square;
        this.flipped = flipped;
        this.size = board.getWidth() / Board.SIZE;
        this.image = Lazy.of(() -> board.getSubimage(left(), top(), size, size));
    }

    public BufferedImage image() {
        return image.get();
    }

    public int size() {
        return size;
    }

    public int left() {
        return size * horizontalOffset(square.col);
    }

    public int right() {
        return size * (horizontalOffset(square.col) + 1);
    }

    public int top() {
        return size * verticalOffset(square.row);
    }

    public int bottom() {
        return size * (verticalOffset(square.row) + 1);
    }

    public int middleX() {
        return left() + size / 2;
    }

    public int middleY() {
        return top() + size / 2;
    }

    private int horizontalOffset(int col) {
        return flipped ? Board.SIZE - 1 - col : col;
    }

    private int verticalOffset(int row) {
        return flipped ? row : Board.SIZE - 1 - row;
    }
}

File: backend/src/main/java/io/chesslave/visual/model/SquareImage.kt
package io.chesslave.visual.model

import io.chesslave.model.Board
import io.chesslave.model.Square
import java.awt.image.BufferedImage

class SquareImage(board: BufferedImage, val square: Square, val flipped: Boolean) {

    val image: BufferedImage by lazy { board.getSubimage(left(), top(), size, size) }

    val size: Int = board.width / Board.SIZE

    fun left(): Int = size * horizontalOffset(square.col)

    fun right(): Int = size * (horizontalOffset(square.col) + 1)

    fun top(): Int = size * verticalOffset(square.row)

    fun bottom(): Int = size * (verticalOffset(square.row) + 1)

    fun middleX(): Int = left() + size / 2

    fun middleY(): Int = top() + size / 2

    private fun horizontalOffset(col: Int): Int = if (flipped) Board.SIZE - 1 - col else col

    private fun verticalOffset(row: Int): Int = if (flipped) row else Board.SIZE - 1 - row

    override fun toString() = square.toString()
}
